### ISSUE TITLE
FSDP2 trained with multi-replica will result in load error during the resume

### ISSUE THREAD

**Issue by StormyXJ** [ByteDance-Seed/VeOmni#172]
When training the model with FSDP2 across multi-node and multi-replica, the DCP can save and resume the checkpoints successfully. However, as I check the first demo image sampled after resuming, it results in noise, different from the result when saving it.
And I try it under single node, the result goes great. Whether there is bug in saving checkpoints with FSDP2 multi-node.

**Comment by Luosuu** [comment]
Hi, what do you mean by "multi-replica"? do you mean HSDP?

**Comment by StormyXJ** [comment]
Yes we are trying to use HSDP, by using the argument named `data_parallel_replicate_size` in the training argument.

**Comment by Luosuu** [comment]
I see. Indeed we did not test much against HSDP. Would you like to share a minimal job yaml that can reproduce your issue? (like what model and do you also use ep?). I will try out later

**Comment by StormyXJ** [comment]
We use Qwen-Image model, and we also find the same issue across different models. We didn't use ep. The following is our config considering the parallel strategy and some training configs.
``` yaml
  data_parallel_mode: fsdp2
  data_parallel_replicate_size: 8 # The one control the num replica
  ulysses_parallel_size: 1
  
  ckpt_manager: dcp

  enable_full_determinism: false
  enable_mixed_precision: true
  enable_gradient_checkpointing: true
  enable_full_shard: true
  enable_fsdp_offload: false
  enable_activation_offload: false
```

In our test, as long as `data_parallel_replicate_size` is bigger than 1, the resumed model performs much worse than it was saved.

**Comment by Luosuu** [comment]
I see. Thank you for reporting. I will investigate this issue later

**Comment by Luosuu** [comment]
Hi @StormyXJ 

Sorry but I could not reproduce your issue

<img width="2508" height="578" alt="Image" src="https://github.com/user-attachments/assets/f4f32629-5c99-41f3-a8f5-d3d7b1d5c9d1" />

could you please tell me more about your job and environment setup? Btw we have switched to PyTorch 2.8 internally for deployment. Please use at least PyTorch 2.7 (never tested with 2.6 for FSDP2)

**Comment by StormyXJ** [comment]
Thanks for your help.
We are using the PyTorch 2.8 too. And I figure it out by referencing [DCP's docs](https://docs.pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict_saver.save), in which it said with HSDP, we only need to save **one of the shard_group**. By doing this, I successfully resume the ckpt.
